libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/Issue.kt

```diff
@@ -12,56 +12,78 @@ sealed interface Issue {
 /**
  * An issue from the GraphQL spec
  */
-sealed interface GraphQLIssue: Issue
+sealed interface GraphQLIssue : Issue
 
 /**
  * A validation issue from the GraphQL spec
  */
-sealed interface GraphQLValidationIssue: GraphQLIssue
+sealed interface GraphQLValidationIssue : GraphQLIssue
 
 /**
  * A custom issue specific to the Apollo compiler
  */
-sealed interface ApolloIssue: Issue
+sealed interface ApolloIssue : Issue
 
 /**
  * A grammar error
  */
-class ParsingError(override val message: String, override val  sourceLocation: SourceLocation?) : GraphQLIssue
+class ParsingError(override val message: String, override val sourceLocation: SourceLocation?) : GraphQLIssue
 
 /**
  * An unknown directive was found.
  *
  * In a perfect world everyone uses SDL schemas, and we can validate directives but in this world, a lot of users rely
  * on introspection schemas that do not contain directives. If this happens, we pass them through without validation.
+ *
+ * In some cases (e.g. `@oneOf`) we want to enforce that the directive is defined. In that case [requireDefinition] is true and the issue
+ * will be raised as an error rather than warning.
  */
-class UnknownDirective(override val message: String, override val sourceLocation: SourceLocation?): GraphQLValidationIssue
+class UnknownDirective(
+    override val message: String,
+    override val sourceLocation: SourceLocation?,
+    val requireDefinition: Boolean,
+) : GraphQLValidationIssue
+
+/**
+ * The directive definition is inconsistent with the expected one.
+ */
+class UnexpectedDirectiveDefinition(
```

Comment:
Nit: Can this be `IncompatibleDirectiveDefinition` ?

Reply:
Haha that was the first name I chose then changed my mind 😅  I'm ok with it too, I'll rename 👍